### PR TITLE
[12.0][ADD] stock_picking_late_activity: new module

### PR DESCRIPTION
**Pull request by ernestotejeda**
This module adds a 'scheduled action' that periodically checks that the 'scheduled date' of the pickings belonging to certain 'operation types' is prior to the datetime that these checks are performed. For each of the pickings that meet this condition, an activity will be created indicating that the 'scheduled date' must be updated.
The operation types that will be taken into account are those that you determine by checking a specific checkbox for that.

Configuration:
Selection_015

Activities for late pickings are atomatically created:
Selection_016

cc @Tecnativa TT25212

### CODE REVIEW

File: stock_picking_late_activity/__manifest__.py [hunk 1]
```diff
@@ -0,0 +1,21 @@
+# Copyright <YEAR(S)> <AUTHOR(S)>
```
Comment (nikul-serpentcs) [inline stock_picking_late_activity/__manifest__.py]:
```suggestion
# Copyright 2020 Tecnativa - Ernesto Tejeda

```

Comment (ernestotejeda) [reply stock_picking_late_activity/__manifest__.py]: Thanks @nikul-serpentcs , change done

File: stock_picking_late_activity/__manifest__.py [hunk 2]
```diff
@@ -0,0 +1,21 @@
+# Copyright 2020 Tecnativa - Ernesto Tejeda
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+{
+    "name": "Stock Picking Late Activity",
+    "summary": "Add an scheduled action that creates late picking activities",
+    "version": "12.0.1.0.0",
+    "development_status": "Stable",
```
Comment (pedrobaeza) [inline stock_picking_late_activity/__manifest__.py]: This development status is not correct:
```suggestion
    "development_status": "Production/Stable",
```

File: stock_picking_late_activity/data/mail_activity_type_data.xml [hunk 3]
```diff
@@ -0,0 +1,12 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!-- Copyright 2020 Tecnativa - Ernesto Tejeda
+     License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl). -->
+<odoo noupdate="1">
+    <record id="mail_activity_type_update_scheduled_date" model="mail.activity.type">
+        <field name="name">Update scheduled date</field>
```
Comment (pedrobaeza) [inline stock_picking_late_activity/data/mail_activity_type_data.xml]:
```suggestion
        <field name="name">Late picking</field>
```

File: stock_picking_late_activity/models/stock_picking.py [hunk 4]
```diff
@@ -0,0 +1,58 @@
+# Copyright 2020 Tecnativa - Ernesto Tejeda
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+
+from datetime import datetime
+from odoo import fields, models
+
+
+class StockPickingType(models.Model):
```
Comment (pedrobaeza) [inline stock_picking_late_activity/models/stock_picking.py]: Put it in a separate file (it is not a one2many model).

File: stock_picking_late_activity/models/stock_picking.py [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+        comodel_name="res.users",
+        string="Late picking activity responsible",
+    )
+
+
+class StockPicking(models.Model):
+    _inherit = "stock.picking"
+
+    def _cron_late_picking_activity(self):
+        """ Create an activity for late pickings. """
```
Comment (pedrobaeza) [inline stock_picking_late_activity/models/stock_picking.py]:
```suggestion
        """Create an activity for late pickings."""
```

File: stock_picking_late_activity/models/stock_picking.py [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+        vals_list = []
+        for picking in late_picking:
+            if not bool(picking.activity_ids.filtered(
+                    lambda r: r.activity_type_id == activity_type)):
+                vals_list.append(picking._prepare_late_picking_activity_vals())
+        self.env["mail.activity"].create(vals_list)
+
+    def _prepare_late_picking_activity_vals(self):
+        """ Hook method to prepare the values to create a late picking
+        activity."""
```
Comment (pedrobaeza) [inline stock_picking_late_activity/models/stock_picking.py]: Follow proper PEP257 for docstring
```suggestion
        activity.
        """
```

File: stock_picking_late_activity/tests/test_stock_picking_late_activity.py [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+        cls.picking_type = cls.env['stock.picking.type'].create({
+            'name': 'Picking type test',
+            'code': 'incoming',
+            'sequence_id': cls.sequence.id,
+            "create_late_picking_activity": True,
+            "late_picking_activity_user_id": cls.user_demo.id,
+        })
+        # Create and validate a picking with 'scheduled_date' == '3 days ago'.
+        scheduled_date = datetime.now() - timedelta(days=3)
+        cls.picking_1 = cls._create_validate_picking(
```
Comment (pedrobaeza) [inline stock_picking_late_activity/tests/test_stock_picking_late_activity.py]: Picking is not validated in the method AFAIK

File: stock_picking_late_activity/tests/test_stock_picking_late_activity.py [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+        self.assertFalse(self.picking_1.activity_ids)
+        # Check no activity for self.picking_3
+        self.assertFalse(self.picking_3.activity_ids)
+        # Check: there is an activity for self.picking_2
+        self.assertEqual(len(self.picking_2.activity_ids), 1)
+        self.assertEqual(
+            self.picking_2.activity_ids.activity_type_id,
+            self.env.ref("stock_picking_late_activity"
+                         ".mail_activity_type_update_scheduled_date"),
+        )
```
Comment (pedrobaeza) [inline stock_picking_late_activity/tests/test_stock_picking_late_activity.py:85]: Check also the user associated to the activity

**Review by pedrobaeza — CHANGES_REQUESTED** [review]

**Comment by pedrobaeza** [comment]
Check Travis

File: stock_picking_late_activity/models/stock_picking.py [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+from odoo import models
+
+
+class StockPicking(models.Model):
+    _inherit = "stock.picking"
+
+    def _cron_late_picking_activity(self):
+        """Create an activity for late pickings."""
+        late_picking = self.search([
+            ("state", 'not in', ["done", "cancel"]),
```
Comment (rousseldenis) [inline stock_picking_late_activity/models/stock_picking.py]: Maybe exclude draft too ?

Comment (ernestotejeda) [reply stock_picking_late_activity/models/stock_picking.py]: done

**Comment by ernestotejeda** [comment]
Changes done

**Review by pedrobaeza — APPROVED** [review]

File: stock_picking_late_activity/readme/CONFIGURE.rst [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+
+#. Go to * Settings > Technical > Automation > Scheduled Actions *
+#. Finds the 'scheduled action' named 'Picking: activities for late pickings'
+#. Edit it and adjust the parameters if you need and save the form.
+
+The late picking activities created will be associated with a new
+'activity type'. If you want to adjust the parameters of that:
+
+#. Go to * Settings> Technical> Email> Activity Types *
+#. Find the 'activity type' named 'Update scheduled date'
```
Comment (bodedra) [inline stock_picking_late_activity/readme/CONFIGURE.rst]:
```suggestion
#. Find the 'activity type' named 'Late picking'
```

**Review by bodedra — APPROVED** [review]
Added minor correction.

Functional reviewed. LGTM :+1:

**Review by marcelsavegnago — APPROVED** [review]

**Comment by ernestotejeda** [comment]
@bodedra changes done

**Pull request merged**